### PR TITLE
[Custom DC] Change import file finder logic

### PR DESCRIPTION
**Pull request by Fructokinase**
Currently files for data import are found by iteratively finding folders and files in gcs. This causes a bug where folders (determined by "/" ending) is not consistently found due to unknown reasons. 

This PR finds all paths with a prefix (root directory of import, also the import group name), and uses a ImportGroupFiles struct to pass data into manifest proto.

### CODE REVIEW

File: gcf/lib/path.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+// CollectImportFiles takes a list of gcs file paths and construct ImportGroupFiles.
+func CollectImportFiles(paths []string) (*ImportGroupFiles, error) {
+
+	dataSource2DatasetNames := map[string][]string{}
+	datasetName2DataFiles := map[string]DataFiles{}
+
+	if len(paths) == 0 {
+		return nil, errors.New("No path found in import group files")
+	}
+	importGroupName := strings.Split(paths[0], "/")[0]
```
Comment (shifucun) [inline gcf/lib/path.go:47]: Add a function comment that all the paths here are under the same import group folder.

File: gcf/lib/path.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	// IMPORTANT NOTE: importGroupName has a character limit of 21
+	importGroupName := strings.Split(paths[0], "/")[0]
+	if len(importGroupName) > 20 {
+		importGroupName = importGroupName[:20]
+	}
+
+	for _, path := range paths {
+		// Path is expected to be like the following:
+		// 	"demo/data/source2/solar/output.csv"
```
Comment (shifucun) [inline gcf/lib/path.go:54]: line54 missing a subfolder?

File: gcf/lib/path.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		// 	"demo/data/source1/smokepm/subfolder/data.tmcf"
+		folderList := strings.SplitN(path, "/", 5)
+		if folderList[0] != importGroupName {
```
Comment (shifucun) [inline gcf/lib/path.go:57]: This is comparing with the truncated version. Should the truncation happen in the end?

File: gcf/lib/path.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+// DataFiles represent a logical unit of files for a single dataset.
+type DataFiles struct {
+	TMCFPath string
+	CSVPaths []string
+}
+
+// ImportGroupFiles represent the collection of paths for a single import group.
+type ImportGroupFiles struct {
+	DataSource2DatasetNames map[string][]string
```
Comment (shifucun) [inline gcf/lib/path.go:32]: nit: no need to have "data" in all the variable names. Can just be "Source2Dataset", "Dataset2Files". It makes the code below easier to read.

File: gcf/custom/manifest.go [hunk 5]
```diff
@@ -22,129 +22,109 @@ import (
...
+		}
+		// skip folders
+		if strings.HasSuffix(attrs.Name, "/") {
+			continue
+		}
+		log.Printf("FindFiles in %s got %s", pathToDataFolder, attrs.Name)
+		paths = append(paths, attrs.Name)
+	}
+
+	// IMPORTANT NOTE: importGroupName has a character limit of 21
```
Comment (shifucun) [inline gcf/custom/manifest.go:118]: importGroupName is not used here?

File: gcf/lib/path.go (hunk 2 above)
Comment (Fructokinase) [reply gcf/lib/path.go:54]: That is intentional, <dataset>/a.tmcf and <dataset>/subfolder/a.tmcf are both valid.

File: gcf/lib/path.go (hunk 1 above)
Comment (Fructokinase) [reply gcf/lib/path.go:47]: done.

File: gcf/lib/path.go (hunk 3 above)
Comment (Fructokinase) [reply gcf/lib/path.go:57]: Thanks for the catch, fixed and added a test case.

File: gcf/lib/path.go (hunk 4 above)
Comment (Fructokinase) [reply gcf/lib/path.go:32]: done.

File: gcf/custom/manifest.go (hunk 5 above)
Comment (Fructokinase) [reply gcf/custom/manifest.go:118]: bad merge, fixed.

File: gcf/lib/path.go (hunk 2 above)
Comment (shifucun) [reply gcf/lib/path.go:51]: oh, actually line 52 "subfolder" is not supported?

**Review by shifucun — DISMISSED** [review]

File: gcf/lib/path.go (hunk 2 above)
Comment (Fructokinase) [reply gcf/lib/path.go:51]: added a constraint.

**Pull request merged**